tern/classes/image_layer.py

```diff
@@ -338,6 +338,45 @@ def gen_fs_hash(self):
             rootfs.extract_tarfile(tar_file, fs_dir)
             self.__fs_hash = rootfs.calc_fs_hash(fs_dir)
 
+    def _parse_hash_content(self, content):
+        """This is an internal function to parse the content of the hash
+        and return a list of FileData objects
+        The content consists of lines of the form:
+        permissions|uid|gid|size|hard links|  sha256sum  filepath xattrs
+        where xattrs is the list of extended attributes for the file
+        The extended attributes start with a '# file' indicator, followed
+        by a list of key-value pairs separated by newlines. For now, we will
+        conserve the key-value pair list as they appear and separate
+        each one by a comma"""
+        file_list = []
+        # keep track of where we are on the list of files
+        index = 0
+        # loop through the content
+        while content:
+            line = content.pop(0)
+            if "# file" in line:
+                # collect the extended attributes
+                xattrlist = []
+                xattrline = content.pop(0)
+                while xattrline != '\n':
+                    xattrlist.append(xattrline.strip())
+                    xattrline = content.pop(0)
+                # when we break out of the extended attributes loop
+                # we combine the results and update the FileData object
+                # existing in the previous index
+                file_list[index-1].extattrs = file_list[index-1].extattrs + \
```

Comment:
what is file_list[-1] for the first iteration of this loop?

Reply:
Also, how does index get updated `if "# file" in line:`? I only see it updated in the `else` statement (maybe this is on purpose?)

Reply:
fs_hash.sh runs `find -type f -printf "%M|%U|%G|%s|%n|` and then `sha256sum` and then `getfattr`. Even if a file is empty, we will still get some results for the permissions and the sha256sum. So my assumption is that the first line will _always_ contain the regular things. Hence, when a `# file` pattern exists in a line it means "we have xattrs for this file but we have already created a file object and placed it at the `index-1` position.

I suppose I can create a file object and then look ahead in `content` at `index+1` to see if there is a `file #` pattern there, but then I would also have to check if the index is within the length of `content` which is iffy because I am changing the length of the content list. So I felt the above assumption makes the processing more efficient.

Does this make sense?

Reply:
Regarding why the index gets bumped in the `else:` part and not the `if` part - it's because this is the index location in `file_list`. For example, I pop line 1 of content. It has all the regular stuff. I make a FileData object, append it to `file_list`, so index 0 in `file_list` is fillled and my next empty spot is at index 1. Now I pop line 2 of content and it has `# file`. That belongs to the FileData object at index 0. Now I have to collect all the extended attributes, make them into something I can attached to the existing attributed in the FileData object at index 0. I am not adding a new FileData object here. I am making a change to the existing FileData object. So I will not be incrementing the index in this case.

Reply:
Thanks for the great explanation. LGTM!